src/persist/src/file.rs

```diff
@@ -0,0 +1,204 @@
+            buf.extend(key_bytes);
+            buf.extend(&val_bytes_len.to_le_bytes());
+            buf.extend(val_bytes);
+            buf.extend(&ts.to_le_bytes());
+            buf.extend(&(*diff as i64).to_le_bytes());
+        }
+        let mut guard = self.dataz.lock()?;
+        guard.write_all(&buf).expect("failed to write file");
+        guard.sync_all().expect("failed to sync file");
+        Ok(())
+    }
+}
+
+impl Meta for FileStream {
+    type Snapshot = MemSnapshot;
+
+    fn snapshot(&self) -> Result<Self::Snapshot, Error> {
+        let mut bytes = Vec::new();
+        let mut dataz = Vec::new();
+
+        let mut guard = self.dataz.lock()?;
+        guard
+            .seek(SeekFrom::Start(0))
+            .expect("failed to seek to beginning");
+        guard.read_to_end(&mut bytes).expect("failed to read file");
+        // Unlock the mutex
+        drop(guard);
+        let mut bytes_remaining = bytes.as_slice();
+
+        while bytes_remaining.len() > 0 {
+            // Decode key
+            let (len_bytes, remainder) = bytes_remaining.split_at(std::mem::size_of::<u64>());
+            let key_len = u64::from_le_bytes(len_bytes.try_into().unwrap());
+
+            let (key_bytes, remainder) = remainder.split_at(key_len as usize);
+            let key = String::from_utf8(key_bytes.to_vec()).expect("failed to parse utf8");
+            bytes_remaining = remainder;
+
+            // Decode val
+            let (len_bytes, remainder) = bytes_remaining.split_at(std::mem::size_of::<u64>());
+            let val_len = u64::from_le_bytes(len_bytes.try_into().unwrap());
+
+            let (val_bytes, remainder) = remainder.split_at(val_len as usize);
+            let val = String::from_utf8(val_bytes.to_vec()).expect("failed to parse utf8");
+            bytes_remaining = remainder;
+
+            // Decode ts and diff
+            let (ts_bytes, remainder) = bytes_remaining.split_at(std::mem::size_of::<u64>());
+            let (diff_bytes, remainder) = remainder.split_at(std::mem::size_of::<i64>());
+            let ts = u64::from_le_bytes(ts_bytes.try_into().unwrap());
+            let diff = i64::from_le_bytes(diff_bytes.try_into().unwrap());
+            bytes_remaining = remainder;
+
+            dataz.push(((key, val), ts, diff as isize));
+        }
+        Ok(MemSnapshot { dataz })
+    }
+
+    fn allow_compaction(&mut self, _ts: u64) -> Result<(), Error> {
+        // No-op for now.
+        Ok(())
+    }
+}
+
+/// Implementation of [Persister]
+pub struct FilePersister {
+    registered: HashSet<Id>,
+    dataz: HashMap<Id, FileStream>,
+}
+
+impl FilePersister {
+    /// Constructs a new, empty SimplePersister.
+    pub fn new() -> Self {
+        FilePersister {
+            registered: HashSet::new(),
+            dataz: HashMap::new(),
+        }
+    }
+
+    /// Consumes this FilePersister, returning the underlying data.
+    /// TODO: I don't really get what these functions are meant to
```

Comment:
We needed them on MemPersister because memory isn't actually persistent and I wanted to simulate a restarted mz process. FilePersister should just take a base path as a constructor argument and remove into_inner and from_inner